v2/pkg/genruntime/extensions/arm_resource_modifier.go

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ */
+
+package extensions
+
+import (
+	"context"
+
+	"github.com/go-logr/logr"
+
+	. "github.com/Azure/azure-service-operator/v2/internal/logging"
+	"github.com/Azure/azure-service-operator/v2/internal/resolver"
+	"github.com/Azure/azure-service-operator/v2/internal/util/kubeclient"
+	"github.com/Azure/azure-service-operator/v2/pkg/genruntime"
+)
+
+// ARMResourceModifier provides a hook allowing resources to modify the payload that will be sent to ARM just before it is sent.
+type ARMResourceModifier interface {
+	// ModifyARMResource takes a genruntime.ARMResource and returns an updated genruntime.ARMResource. The updated resource
+	// is then serialized and sent to ARM in the body of a PUT request.
+	ModifyARMResource(
+		ctx context.Context,
+		armObj genruntime.ARMResource,
+		obj genruntime.MetaObject,
+		kubeClient kubeclient.Client,
+		resolver *resolver.Resolver,
+		log logr.Logger) (genruntime.ARMResource, error)
+}
+
+type ARMResourceModifierFunc = func(ctx context.Context, obj genruntime.MetaObject, armObj genruntime.ARMResource) (genruntime.ARMResource, error)
+
+// CreateARMResourceModifier returns a function that performs per-resource modifications. If the ARMResourceModifier extension has
+// not been implemented for the resource in question, the default behavior is to return the provided genruntime.ARMResource unmodified.
+func CreateARMResourceModifier(
+	host genruntime.ResourceExtension,
+	kubeClient kubeclient.Client,
+	resolver *resolver.Resolver,
+	log logr.Logger) ARMResourceModifierFunc {
+
+	impl, ok := host.(ARMResourceModifier)
+	if !ok {
+		return func(ctx context.Context, obj genruntime.MetaObject, armObj genruntime.ARMResource) (genruntime.ARMResource, error) {
+			return armObj, nil
+		}
+	}
+
+	return func(ctx context.Context, obj genruntime.MetaObject, armObj genruntime.ARMResource) (genruntime.ARMResource, error) {
+		log.V(Info).Info("Modifying ARM payload")
+
+		armResource, err := impl.ModifyARMResource(ctx, armObj, obj, kubeClient, resolver, log)
+		if err != nil {
+			return nil, err
```

Comment:
Fixed the missing log for the error path. Asked a question on GVK out of band, will come back around to this based on that.

Reply:
As discussed, the logs should contain enough information on this already since log messages always contain the controller that hit the issue (from which you can derive the GVK)